src/homework4.py

```diff
@@ -0,0 +1,421 @@
+        cur_response = response[i]
+
+        for bin_num in range(num_bins):
+            if (cat_check is True and cur_val == possible_cat_predictors[bin_num]) or (
+                cat_check is False
+                and cur_val >= lower_bins[bin_num]
+                and cur_val <= upper_bins[bin_num]
+            ):
+                bin_counts[bin_num] += 1
+                bin_responses[bin_num] += cur_response
+                break
+
+    population_mean = np.mean(response)
+
+    for i in range(num_bins):
+        bin_mean = bin_responses[i] / bin_counts[i]
+        bin_means.append(bin_mean)
+        bin_mean_minus_pop_mean.append(bin_mean - population_mean)
+        mean_squared_diff.append((bin_mean - population_mean) ** 2)
+
+        weight = bin_counts[i] / len(predictor)
+        weighted_mean_squared_diff.append(
+            (weight * ((bin_mean - population_mean) ** 2))
+        )
+
+    bin_mean_minus_pop_mean = np.array(bin_mean_minus_pop_mean)
+
+    fig = make_subplots(specs=[[{"secondary_y": True}]])
+
+    if cat_check is True:
+        fig.add_trace(
+            go.Bar(x=possible_cat_predictors, y=bin_counts, name="Population"),
+            secondary_y=False,
+        )
+
+        possible_cat_predictors.sort()
+
+        fig.add_trace(
+            go.Scatter(
+                x=possible_cat_predictors,
+                y=bin_mean_minus_pop_mean,
+                name="\u03BC(i) - \u03BC(population)",
+            ),
+            secondary_y=True,
+        )
+    else:
+        fig.add_trace(
+            go.Bar(x=bin_centers, y=bin_counts, name="Population"), secondary_y=False
+        )
+
+        fig.add_trace(
+            go.Scatter(
+                x=bin_centers,
+                y=bin_mean_minus_pop_mean,
+                name="\u03BC(i) - \u03BC(population)",
+            ),
+            secondary_y=True,
+        )
+
+    fig.add_hline(y=population_mean, name="population mean")
+
+    fig.update_layout(
+        title_text="Difference with Mean of Response - Predictor {}".format(
+            predictor_name
+        )
+    )
+
+    fig.update_xaxes(title_text="Predictor Bin")
+    fig.update_yaxes(title_text="Response", secondary_y=False)
+    fig.update_yaxes(
+        title_text="Population",
+        secondary_y=True,
+        range=[-(max(response)), max(response)],
+    )
+    fig.update_layout(
+        dict(
+            yaxis2={"anchor": "x", "overlaying": "y", "side": "left"},
+            yaxis={"anchor": "x", "domain": [0.0, 1.0], "side": "right"},
+        )
+    )
+    fig.show()
+
+    dif_w_mean_of_resp = np.mean(mean_squared_diff)
+    print(
+        "Difference with mean of response (unweighted) = {}".format(
+            str(dif_w_mean_of_resp)
+        )
+    )
+    print(
+        "Difference with mean of response (weighted) = {}".format(
+            str(sum(weighted_mean_squared_diff))
+        )
+    )
+
+
+def random_forest(df, cont_list, response):
+    cont_list.append(response)
```

Comment:
random forest function looks good! I forgot to compare predictors together for myself, and it was wrong for me; glad that you made this right!

Reply:
however, it's a bit confusing to me that you're appending response into the list, when it's supposed to be adding predictor. Not sure if that would cause any trouble.